org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/util/SemanticHighlightingTokens.xtend

```diff
@@ -0,0 +1,172 @@
+		if (tokens.nullOrEmpty) {
+			return '';
+		}
+		// 2 stands for: number of elements for the output; the "character" and the "length and scope".
+		// 4 stands for: 4 byte for a primitive integer.
+		val buffer = ByteBuffer.allocate(tokens.size * 2 * 4);
+		for (token : tokens) {
+			val character = token.character;
+			val length = token.length;
+			val scope = token.scope;
+
+			var lengthAndScope = length;
+			lengthAndScope = lengthAndScope << LENGTH_SHIFT;
+			lengthAndScope = lengthAndScope.bitwiseOr(scope);
+
+			buffer.putInt(character);
+			buffer.putInt(lengthAndScope);
+		}
+		return Base64.getEncoder().encodeToString(buffer.array());
+	}
+
+	/**
+	 * Decodes the tokens string and returns with a list of semantic highlighting tokens.
+	 * Returns with an empty list if the argument is {@code null} or empty.
+	 */
+	static def List<Token> decode(String tokens) {
+		if (tokens === null || tokens.length === 0) {
+			return emptyList;
+		}
+		val buffer = ByteBuffer.wrap(Base64.getDecoder().decode(tokens));
+		val builder = ImmutableList.builder;
+		while (buffer.hasRemaining()) {
+			val character = buffer.getInt();
+
+			val lengthAndScope = buffer.getInt();
+			val length = lengthAndScope >> LENGTH_SHIFT;
+			val scope = lengthAndScope.bitwiseAnd(SCOPES_MASK);
+
+			builder.add(new Token(
+				character,
+				length,
+				scope
+			));
+		}
+		return builder.build;
+	}
+
+	/**
+	 * The bare minimum representation of a semantic highlighting token.
+	 * 
+	 * <p>
+	 * Semantic highlighting tokens are <b>not</b> part of the protocol, as they're
+	 * sent through the wire as an encoded string. The purpose of this data type is
+	 * to help working with them.
+	 */
+	@Data
+	static class Token implements Comparable<Token> {
+		
+		static val EMPTY_ARRAY = newIntArrayOfSize(0);
+
+		/**
+		 * Converts an iterable of tokens into an array of primitive integers. Returns with
+		 * an empty array if the argument is {@code null} or empty.
+		 * 
+		 * <p>
+		 * <ul>
+		 *   <li>{@code array[3*i]} is the {@link Token#character character} of the token.</li>
+		 *   <li>{@code array[3*i+1]} is the {@link Token#length length} of the token.</li>
+		 *   <li>{@code array[3*i]} is the token {@link Token#scope scope}.</li>
+		 * </ul>
+		 */
+		static def toIntArray(Iterable<? extends Token> tokens) {
+			if (tokens.nullOrEmpty) {
+				return EMPTY_ARRAY;
+			}
+			val array = newIntArrayOfSize(tokens.size * 3);
+			var i = 0;
+			for (token : tokens) {
+				array.set(i++, token.character);
+				array.set(i++, token.length);
+				array.set(i++, token.scope);
+			}
+			return array;
+		}
+
+		/**
+		 * Counter-part of the {@link Token#toIntArray}. Converts an array of primitive integers
+		 * into a list of tokens. If the input is {@code null}, returns with an empty list.
+		 * Throws an exception if the length of the arguments is not a modulo of {@code 3}.
+		 */
+		static def fromIntArray(int... input) {
```

Comment:
I need to an assertion here. The ints cannot exceed the 0 and 2<sup>32</sup>-1 (inclusive) range.